src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs

```diff
@@ -141,54 +141,98 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
 
             try
             {
-                await ProcessMessage(eventArgs).ConfigureAwait(false);
+                await Process(eventArgs).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                Logger.Warn("Failed to process message. Returning message to queue...", ex);
+                await ReturnToQueue(eventArgs.DeliveryTag).ConfigureAwait(false);
             }
             finally
             {
                 semaphore.Release();
             }
         }
 
-        async Task ProcessMessage(BasicDeliverEventArgs message)
+        async Task Process(BasicDeliverEventArgs message)
         {
-            string messageId = null;
+            string messageId;
+            try
+            {
+                messageId = messageConverter.RetrieveMessageId(message);
+            }
+            catch (Exception ex)
+            {
+                Logger.Error($"Failed to retreive ID from poison message. Moving message to queue '{settings.ErrorQueue}'...", ex);
```

Comment:
I went with "Moving" over "Forwarding" since we tend to favour that language, e.g.

> Our message-processing infrastructure must *move* these poison messages to the side...

>...it would make sense to create a single centralized error queue to which all poison messages could be *moved*.

<sup>http://particular.net/blog/but-all-my-errors-are-severe</sup>

> This sort of failure case is best handled by having the messaging infrastructure just *move* the message to an error queue.

<sup>http://udidahan.com/2007/02/15/its-poison-i-tell-you-poison/</sup>